File: Solutions/1306 Jump Game III/1306 Jump Game III.js
/**
 * @param {number[]} arr
 * @param {number} start
 * @return {boolean}
 */
var canReach = function(arr, start) {

    function callDFS(idx) {
        if(arr[idx] === 'seen' || idx < 0 || idx >= arr.length) return false;
        if(!arr[idx]) return true;
        const move = arr[idx]; arr[idx] = 'seen';
        return callDFS(idx - move) || callDFS(idx + move);
    }
    return callDFS(start);
};

//Third fastest
var canReach = function(arr, start) {
    const visited = new Set();
    return jump(arr, start);
    
    function jump(arr, start) {
        const n = arr.length;
        if (start < 0 || start >= n) return false;
        if (arr[start] === 0) return true;
        if (visited.has(start)) {
            return false;
        } else {
            visited.add(start);
        }

        return jump(arr, start + arr[start]) || jump(arr, start - arr[start]);
    }
};


//Second fastest
var canReach = function(arr, start, indexes = []) {
    
    if(arr[start] === 0)
        return true;
    else if(indexes.includes(start))
        return false;
    
    indexes.push(start);
    
    let newStart = start - arr[start];    
    if(newStart >= 0)
        if(canReach(arr, newStart, indexes))
            return true;
    
    newStart = start + arr[start];
    if(newStart < arr.length)
        if(canReach(arr, newStart, indexes))
            return true;
    
    return false;
    
};

//Fastest
var canReach = function(arr, start) {
    function find0(input, i) {
        let visited = {};
        function jump(input, jumpTo) {
        //console.log(jumpTo);
        if (input[jumpTo] == 0) return true;
        if (jumpTo < 0 || jumpTo > input.length) return false;
        if (visited[jumpTo]) return false;

        visited[jumpTo] = true;

        return (
             jump(input, jumpTo + input[jumpTo]) || jump(input, jumpTo - input[jumpTo])
        );
  }

  return jump(input, i);
    }
    
    return find0(arr, start)
};

File: Solutions/1306 Jump Game III/1306 Jump Game III.go
//Third fastest (Most common)
func canReach(arr []int, start int) bool {
    visited:=make(map[int]bool)

    return jump(arr, start + arr[start], visited) || jump(arr, start - arr[start], visited)
}

func jump(arr []int, next int, visited map[int]bool) bool {
    if next < 0 || next >= len(arr) {
        return false
    }
    
    if arr[next] == 0 {
        return true
    }
    
    if visited[next] {
        return false
    }
    
    visited[next] = true
    
    return jump(arr, next + arr[next], visited) || jump(arr, next - arr[next], visited)
    
}

//Second fastest
var visited []bool

func canReach(arr []int, start int) bool {
	visited = make([]bool, len(arr))
	return canReachRec(arr, start)
}

func canReachRec(arr []int, start int) bool {
	if start >= len(arr) || start < 0 {
		return false
	}
	if visited[start] {
		return false
	}
	if arr[start] == 0 {
		return true
	}
	visited[start] = true
	left := start - arr[start]
	right := start + arr[start]

	return canReachRec(arr, right) || canReachRec(arr, left)
}

//Fastest
func canReach(arr []int, s int) bool {
    moves := []int{}
    visited := map[int]bool{s:true}
    if m := s+arr[s]; m < len(arr){
        moves = append(moves, m)
        visited[m] = true
    }
    if m := s-arr[s]; m >= 0{
        moves = append(moves, m)
        visited[m] = true
    }
    for len(moves) > 0{
        nextmoves := []int{}
        for _, m := range moves{
            if arr[m] == 0{return true}
            if newm := m+arr[m]; newm < len(arr) && !visited[newm]{
                nextmoves = append(nextmoves, newm)
                visited[newm] = true
            }
            if newm := m-arr[m]; newm >= 0 && !visited[newm]{
                visited[newm] = true
                nextmoves = append(nextmoves, newm)
            }
        }
        moves = nextmoves
    }
    return false
}